Panic: index out of bounds while parsing

I get an out of bounds panic while parsing this particular input.:
```zig
test "test arocc parse" {
    var diagnostics: aro.Diagnostics = .{ .output = .ignore };
    var comp = aro.Compilation.init(std.testing.allocator, &diagnostics, std.fs.cwd());
    defer comp.deinit();

    const file = try comp.addSourceFromBuffer("file.c",
        \\int foo = 5;
        \\int bar = 100;
        \\
        \\int *someFunc(int x, int y) { return &foo; }
        \\
        \\struct Foo {
        \\   int x: 3;
        \\   int y;
        \\   int z;
        \\};
    );

    var pp = aro.Preprocessor.init(&comp, .default);
    defer pp.deinit();
    try pp.addBuiltinMacros();

    const builtin_macros = try comp.generateBuiltinMacros(.no_system_defines);
    _ = try pp.preprocess(builtin_macros);

    const eof = try pp.preprocess(file);
    try pp.addToken(eof);

    var tree = try aro.Parser.parse(&pp);
    defer tree.deinit();
}
```


Error output/stack trace:
```
thread 72809 panic: index out of bounds: index 43, len 43
/home/ronald/.cache/zig/p/aro-0.0.0-JSD1QpUCJwDojnxIuz2DILgM8e5YFO4G_hDihrst3gqm/src/aro/Parser.zig:989:26: 0x118c79a in nextExternDecl (aro.zig)
        switch (p.tok_ids[p.tok_i]) {
                         ^
/home/ronald/.cache/zig/p/aro-0.0.0-JSD1QpUCJwDojnxIuz2DILgM8e5YFO4G_hDihrst3gqm/src/aro/Parser.zig:915:25: 0x112a78c in parse (aro.zig)
        p.nextExternDecl();
                        ^
/home/ronald/proj/bfd-zig/src/main.zig:227:36: 0x10f148e in test.test arocc parse (main.zig)
    var tree = try aro.Parser.parse(&pp);
                                   ^
/home/ronald/opt/zig/lib/compiler/test_runner.zig:128:29: 0x1789bc1 in mainServer (test_runner.zig)
                test_fn.func() catch |err| switch (err) {
                            ^
/home/ronald/opt/zig/lib/compiler/test_runner.zig:62:26: 0x178ae80 in main (test_runner.zig)
        return mainServer() catch @panic("internal test runner failure");
                         ^
/home/ronald/opt/zig/lib/std/start.zig:666:22: 0x1784809 in main (std.zig)
            root.main();
                     ^
../sysdeps/nptl/libc_start_call_main.h:58:16: 0x7950e1a2a1c9 in __libc_start_call_main (../sysdeps/x86/libc-start.c)
../csu/libc-start.c:360:3: 0x7950e1a2a28a in __libc_start_main_impl (../sysdeps/x86/libc-start.c)
???:?:?: 0x17b8ba4 in ??? (???)
???:?:?: 0x0 in ??? (???)


```

This reproduces on master if invoking aro using the `-undef` flag:
```
ehaas@Mac ~ % arocc -fsyntax-only  crash.c
ehaas@Mac ~ % arocc -fsyntax-only -undef crash.c
crash.c:1:1: error: expected external declaration
int foo = 5;
^
panic: index out of bounds: index 43, len 43
/Users/ehaas/source/arocc/src/aro/Parser.zig:989:26: 0x1049f845f in nextExternDecl (arocc)
        switch (p.tok_ids[p.tok_i]) {
                         ^
/Users/ehaas/source/arocc/src/aro/Parser.zig:915:25: 0x1049aaee3 in parse (arocc)
        p.nextExternDecl();
                        ^
/Users/ehaas/source/arocc/src/aro/Driver.zig:924:28: 0x1049a5d1b in processSource__anon_28929 (arocc)
    var tree = try pp.parse();
                           ^
/Users/ehaas/source/arocc/src/aro/Driver.zig:794:28: 0x104910a9f in main__anon_24681 (arocc)
        try d.processSource(tc, source, builtin_macros, user_macros, fast_exit, asm_gen_fn);
                           ^
/Users/ehaas/source/arocc/src/main.zig:65:16: 0x10490bfcf in main (arocc)
    driver.main(&toolchain, args, fast_exit, assembly_backend.genAsm) catch |er| switch (er) {
               ^
/Users/ehaas/local/zig-macos-aarch64-0.15.0-dev.471+369177f0b/lib/std/start.zig:666:29: 0x10490b937 in main (arocc)
            return root.main();
                            ^
???:?:?: 0x19c6beb97 in ??? (???)
???:?:?: 0x0 in ??? (???)
```

--- a/src/aro/Parser.zig
+++ b/src/aro/Parser.zig
@@ -803,6 +803,9 @@ pub fn parse(pp: *Preprocessor) Compilation.Error!Tree {
     assert(pp.linemarkers == .none);
     pp.comp.pragmaEvent(.before_parse);
 
+    const expected_implicit_typedef_max = 7;
+    try pp.tokens.ensureUnusedCapacity(pp.gpa, expected_implicit_typedef_max);
+
     var p: Parser = .{
         .pp = pp,
         .comp = pp.comp,
@@ -810,7 +813,7 @@ pub fn parse(pp: *Preprocessor) Compilation.Error!Tree {
         .gpa = pp.comp.gpa,
         .tree = .{
             .comp = pp.comp,
-            .tokens = pp.tokens.slice(),
+            .tokens = undefined, // Set after implicit typedefs
         },
         .tok_ids = pp.tokens.items(.id),
         .strings = .init(pp.comp.gpa),
@@ -866,8 +869,12 @@ pub fn parse(pp: *Preprocessor) Compilation.Error!Tree {
         if (p.comp.float80Type()) |float80_ty| {
             try p.addImplicitTypedef("__float80", float80_ty);
         }
+
+        // Set here so that the newly generated tokens are included.
+        p.tree.tokens = p.pp.tokens.slice();
     }
     const implicit_typedef_count = p.decl_buf.items.len;
+    assert(implicit_typedef_count <= expected_implicit_typedef_max);
 
     while (p.eatToken(.eof) == null) {
         if (try p.pragma()) continue;
@@ -935,15 +942,12 @@ fn addImplicitTypedef(p: *Parser, name: []const u8, qt: QualType) !void {
     try p.comp.generated_buf.append(p.comp.gpa, '\n');
 
     const name_tok: u32 = @intCast(p.pp.tokens.len);
-    try p.pp.tokens.append(p.gpa, .{ .id = .identifier, .loc = .{
+    p.pp.tokens.appendAssumeCapacity(.{ .id = .identifier, .loc = .{
         .id = .generated,
         .byte_offset = @intCast(start),
         .line = p.pp.generated_line,
     } });
     p.pp.generated_line += 1;
-
-    // Reset in case there was an allocation.
-    p.tree.tokens = p.pp.tokens.slice();
 
     const node = try p.addNode(.{
         .typedef = .{

--- a/test/cases/undef.c
+++ b/test/cases/undef.c
@@ -3,3 +3,16 @@
 #if defined(linux) || defined(__linux) || defined(__linux__)
 #error Should not be defined
 #endif
+
+int foo = __CHAR_BIT__;
+int bar = 100;
+
+int *someFunc(int x, int y) { return &bar; }
+
+struct Foo {
+   int x: 3;
+   int y;
+   int z;
+};
+
+#define EXPECTED_ERRORS "undef.c:7:11: error: use of undeclared identifier '__CHAR_BIT__'"

